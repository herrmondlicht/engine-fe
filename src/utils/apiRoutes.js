import * as axios from "axios";

export const requestMaker = (
  { method, url, isLocked },
  { storage = window.localStorage } = {}
) => ({ data, params }) => {
  const headers = {};
  if (isLocked) headers.Authorization = storage.getItem("token");
  return axios({
    headers,
    method,
    url: `${process.env.REACT_APP_API_SERVICE}/api${url}`,
    data,
    params
  });
};

const routes = {
  login: {
    url: "/session/login",
    isLocked: false
  }
};

const routesWithRequests = Object.keys(routes).reduce(
  (routesReduced, routeKey) => ({
    ...routesReduced,
    [routeKey]: {
      get: requestMaker({
        method: "get",
        isLocked: routes[routeKey].isLocked,
        url: routes[routeKey].url
      }),
      post: requestMaker({
        method: "post",
        isLocked: routes[routeKey].isLocked,
        url: routes[routeKey].url
      })
    }
  }),
  {}
);

export default routesWithRequests;
