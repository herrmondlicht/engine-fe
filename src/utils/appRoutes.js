import AddNewCar from "../components/CarAndCustomerForm/AddCarAndCustomer";

const routes = {
  addnewcar: {
    name: "Add new car",
    path: "addnewcar",
    Component: AddNewCar,
    isLocked: true,
    exact: true
  },
  home: {
    name: "Add new car",
    path: "/",
    Component: AddNewCar,
    isLocked: true,
    exact: true
  }
};

export default routes;
