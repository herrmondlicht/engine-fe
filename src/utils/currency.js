const NOT_DIGIT = /\D/g;

export const toBRL = value => {
  if (isNaN(value)) {
    return "-";
  }
  return Number(value.toString()).toLocaleString("pt-br", {
    style: "currency",
    currency: "BRL",
  });
};

export const fromBRL = value => {
  if (value === undefined || value === null) {
    return "-";
  }
  return fixDecimalPoint(value.toString());
};

export const handleCurrencyFieldChange = value => {
  const fixedValue = fixDecimalPoint(value.toString()).split("");
  if (fixedValue[0] === ".") {
    fixedValue.splice(0, 0, "0");
  }
  if (fixedValue[0] === "0" && fixedValue.length > 4) {
    fixedValue.shift();
  }
  return toBRL(fixedValue.join(""));
};

export const sanitizeNumber = value => value?.replace(NOT_DIGIT, "");

const fixDecimalPoint = value => {
  if (!value) {
    return "0.00";
  }
  const sanitizedValueRaw = sanitizeNumber(value.toString());
  const valueRawArray = sanitizedValueRaw.split("");
  valueRawArray.splice(-2, 0, ".");
  const valueWithDecimalPoint = valueRawArray.join("");

  return valueWithDecimalPoint;
};
