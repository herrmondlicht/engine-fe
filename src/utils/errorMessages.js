export const getErrorMessage = (errorType) => {
  switch (errorType) {
  case "":
    return "";
  case "CAR_FORM":
    return "Não foi possível adicionar o carro. Verifique os dados e tente novamente";
  case "CUSTOMER_FORM":
    return "Não foi possível adicionar o cliente. Verifique os dados e tente novamente";
  case "CUSTOMER_CAR_FORM":
    return "Não foi possível adicionar o carro do cliente. Verifique os dados e tente novamente";
  default:
    return "Não foi possível completar a ação";
  }
};
