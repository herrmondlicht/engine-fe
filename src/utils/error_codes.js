export const ERROR_CODES = {
  DUP00001: "DUP00001",
  AUTH00001: "AUTH00001",
  INTERNAL_ERROR: "INTERNAL_ERROR",
};

export const MAPPED_ERROR = {
  ER_DUP_ENTRY: {
    status: 400,
    code: ERROR_CODES.DUP00001,
    error: "Duplicated object in database.",
  },
  UnauthorizedError: {
    status: 401,
    code: ERROR_CODES.AUTH00001,
    error: "Invalid token",
  },
  invalid_token: {
    status: 401,
    code: ERROR_CODES.AUTH00001,
    error: "Invalid token",
  },
  INTERNAL_ERROR: {
    status: 500,
    code: ERROR_CODES.INTERNAL_ERROR,
    error: "Something went wrong",
  },
};
