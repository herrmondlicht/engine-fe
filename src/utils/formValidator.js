import * as yup from "yup";

yup.setLocale({
  mixed: {
    default: "Não é válido",
    required: "Campo obrigatório",
  },
});

export { yup };
