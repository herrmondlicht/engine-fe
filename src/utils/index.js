export * from "./errorMessages";
export * from "./storage/storageAPI";
export * from "./engineAPI/engineAPI";
export * from "./appRoutes";