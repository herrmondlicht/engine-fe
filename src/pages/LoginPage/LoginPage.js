import React, { useEffect, useRef } from "react";
import { storageAPI, STORAGE_KEYS, engineAPI, yup } from "utils";
import { useLocation, useHistory, Redirect } from "react-router-dom";
import EngineImage from "assets/engine_logo.png";
import { LoginForm } from "components";
import { Card } from "ui-fragments";
import { useCustomForm, useLoader, useNotification } from "hooks";

const loginFormSchema = yup.object().shape({
  password: yup.string().required(),
  username: yup.string().required(),
});

export const createLogin = (engineAPI, storageAPI) =>
  function Login() {
    const { showErrorNotification } = useNotification();
    const [isLoading, setIsLoading] = useLoader(false);
    const {
      formMethods: { handleSubmit, reset, register },
      validationMethods: { validate, errors },
    } = useCustomForm({ schema: loginFormSchema });
    const isMounted = useRef(true);
    const location = useLocation();
    const history = useHistory();
    const token = storageAPI.getItem(STORAGE_KEYS.TOKEN);
    const { from } = location.state || { from: { pathname: "/" } };

    useEffect(
      () => () => {
        isMounted.current = false;
      },
      []
    );

    async function sendForm(loginData) {
      if (!validate(loginData)) {
        return;
      }
      try {
        setIsLoading(true);
        const data = await engineAPI.login.post({ data: loginData });
        loginUser(data.token);
      } catch (e) {
        // TODO: implement proper error handling
        if (e?.response?.status === 401) {
          showErrorNotification({
            id: "authError",
            message: "Usuário ou senha incorretos",
          });
        } else {
          showErrorNotification({
            id: "loginError",
            message: "Não conseguimos acessar o sistema!",
          });
        }
      } finally {
        setIsLoading(false);
      }

      isMounted.current && setIsLoading(false);
    }

    function loginUser(token) {
      storageAPI.setItem(STORAGE_KEYS.TOKEN, token);
      reset();
      history.replace(from);
    }

    if (token) {
      return <Redirect to={{ pathname: from.pathname }} />;
    }

    return (
      <div
        data-testid="LoginPage"
        className="flex w-full h-full justify-center items-center relative"
      >
        <Card>
          <img src={EngineImage} alt="logo" className="w-80 mb-10" />
          <LoginForm
            registerInput={register}
            sendForm={handleSubmit(sendForm)}
            isLoading={isLoading}
            errors={errors}
          />
        </Card>
      </div>
    );
  };

export default createLogin(engineAPI, storageAPI);
