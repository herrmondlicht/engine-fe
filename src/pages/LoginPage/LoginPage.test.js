import React from "react";
import { stub, assert } from "sinon";
import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { createLogin } from "./LoginPage";
import { STORAGE_KEYS } from "utils";

describe("Login Page", () => {
  it("should render", () => {
    const engineAPIStub = { login: { post: stub() } };
    const LoginPage = createLogin(engineAPIStub, {
      getItem: stub(),
      setItem: stub(),
    });

    render(<LoginPage />, { wrapper: BrowserRouter });

    expect(screen.getByTestId("LoginPage")).toMatchSnapshot();
  });

  it("should display error when username is empty and user clicks login", async () => {
    const engineAPIStub = { login: { post: stub() } };
    const LoginPage = createLogin(engineAPIStub, {
      getItem: stub(),
      setItem: stub(),
    });

    render(<LoginPage />, { wrapper: BrowserRouter });

    screen.getByText("Entrar").click();
    await waitFor(() => {
      const inputUser = screen.getByLabelText("Usuário");
      within(inputUser.parentElement).getByText("Campo obrigatório");
    });
  });

  it("should display error when password is empty and user clicks login", async () => {
    const engineAPIStub = { login: { post: stub() } };
    const LoginPage = createLogin(engineAPIStub, {
      getItem: stub(),
      setItem: stub(),
    });

    render(<LoginPage />, { wrapper: BrowserRouter });
    const inputUser = screen.getByLabelText("Usuário");
    const inputPassword = screen.getByLabelText("Senha");

    fireEvent.change(inputUser, { target: { value: "admin" } });
    screen.getByText("Entrar").click();

    within(inputPassword.parentElement).findByText("Campo obrigatório");
  });

  it("should call api post method with correct data", async () => {
    const engineAPIStub = {
      login: { post: stub().resolves({ token: "abce" }) },
    };
    const sessionStorageStub = {
      getItem: stub(),
      setItem: stub(),
    };
    const LoginPage = createLogin(engineAPIStub, sessionStorageStub);
    const EXPECTED_CALL_ARG = {
      data: {
        username: "admin",
        password: "password",
      },
    };

    render(<LoginPage />, { wrapper: BrowserRouter });
    const inputUser = screen.getByLabelText("Usuário");
    const inputPassword = screen.getByLabelText("Senha");

    fireEvent.change(inputUser, { target: { value: "admin" } });
    fireEvent.change(inputPassword, { target: { value: "password" } });
    screen.getByText("Entrar").click();

    await waitFor(() => {
      assert.calledWithExactly(engineAPIStub.login.post, EXPECTED_CALL_ARG);
      assert.calledWithExactly(
        sessionStorageStub.setItem,
        STORAGE_KEYS.TOKEN,
        "abce"
      );
    });
  });
});
