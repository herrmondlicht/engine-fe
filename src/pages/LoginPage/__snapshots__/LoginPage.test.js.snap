// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render 1`] = `
<div
  class="flex w-full h-full justify-center items-center relative"
  data-testid="LoginPage"
>
  <div
    class="p-4 md:p-9 shadow-md rounded-3xl bg-white "
  >
    <img
      alt="logo"
      class="w-80 mb-10"
      src="engine_logo.png"
    />
    <form
      class="flex flex-col gap-3 w-full"
      data-testid="LoginFormContainer"
    >
      <div
        class="w-full"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="mb-4 ml-1"
          >
            <label
              class="text-md text-gray-500"
              id="username-label"
            >
              Usu√°rio
            </label>
          </div>
          <input
            aria-labelledby="username-label"
            autocomplete="no"
            class="p-4 bg-gray-100 placeholder-gray-400 rounded-2xl text-gray-800 focus:bg-white focus:ring-2 focus:ring-primary-0 resize-none w-full css-1bya8sb css-9bnagd"
            name="username"
          />
        </div>
      </div>
      <div
        class="w-full"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="mb-4 ml-1"
          >
            <label
              class="text-md text-gray-500"
              id="password-label"
            >
              Senha
            </label>
          </div>
          <input
            aria-labelledby="password-label"
            autocomplete="no"
            class="p-4 bg-gray-100 placeholder-gray-400 rounded-2xl text-gray-800 focus:bg-white focus:ring-2 focus:ring-primary-0 resize-none w-full css-1bya8sb css-9bnagd"
            name="password"
            type="password"
          />
        </div>
      </div>
      <div
        class="mt-4"
      >
        <button
          class="p-3 px-5 text-md bg-primary-0 focus:ring-primary-0 hover:bg-primary-1 focus:ring-2 focus:border-transparent border-2 text-white w-full rounded-full font-semibold transition duration-300"
        >
          <div
            class="relative w-full h-full flex justify-center items-center"
          >
            <div
              class="transition duration-400 opacity-100"
            >
              Entrar
            </div>
            <div
              class="absolute z-10 transition duration-400 opacity-0"
            >
              <div
                aria-busy="true"
                class=""
              >
                <svg
                  aria-label="audio-loading"
                  height="30"
                  viewBox="0 0 38 38"
                  width="30"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs>
                    <lineargradient
                      id="a"
                      x1="8.042%"
                      x2="65.682%"
                      y1="0%"
                      y2="23.865%"
                    >
                      <stop
                        offset="0%"
                        stop-color="#fff"
                        stop-opacity="0"
                      />
                      <stop
                        offset="63.146%"
                        stop-color="#fff"
                        stop-opacity=".631"
                      />
                      <stop
                        offset="100%"
                        stop-color="#fff"
                      />
                    </lineargradient>
                  </defs>
                  <g
                    fill="none"
                    fill-rule="evenodd"
                  >
                    <g
                      transform="translate(1 1)"
                    >
                      <path
                        d="M36 18c0-9.94-8.06-18-18-18"
                        id="Oval-2"
                        stroke="#fff"
                        stroke-width="2"
                      >
                        <animatetransform
                          attributeName="transform"
                          dur="0.9s"
                          from="0 18 18"
                          repeatCount="indefinite"
                          to="360 18 18"
                          type="rotate"
                        />
                      </path>
                      <circle
                        cx="36"
                        cy="18"
                        fill="#fff"
                        r="1"
                      >
                        <animatetransform
                          attributeName="transform"
                          dur="0.9s"
                          from="0 18 18"
                          repeatCount="indefinite"
                          to="360 18 18"
                          type="rotate"
                        />
                      </circle>
                    </g>
                  </g>
                </svg>
              </div>
            </div>
            <div
              class="absolute z-10 transition duration-400 opacity-0"
            />
          </div>
        </button>
      </div>
    </form>
  </div>
</div>
`;
